File: solutions/cpp/Points_covering_intervals_alternative.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <deque>
#include <iostream>
#include <random>
#include <set>
#include <unordered_set>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::deque;
using std::endl;
using std::max;
using std::random_device;
using std::uniform_int_distribution;
using std::unordered_set;
using std::vector;

struct EndPoint;
vector<int> FindMinimumVisitsHelper(const vector<EndPoint>& endpoints);

// @include
struct Interval {
  int left, right;
};

struct EndPoint {
  bool operator<(const EndPoint& that) const {
    const int a = is_left ? ptr->left : ptr->right,
              b = that.is_left ? that.ptr->left : that.ptr->right;
    return a < b;
  }

  const Interval* ptr;
  bool is_left;
};

vector<int> FindMinimumVisits(const vector<Interval>& I) {
  vector<EndPoint> endpoints;
  for (size_t i = 0; i < I.size(); ++i) {
    endpoints.emplace_back(EndPoint{&I[i], true});
    endpoints.emplace_back(EndPoint{&I[i], false});
  }
  sort(endpoints.begin(), endpoints.end());

  return FindMinimumVisitsHelper(endpoints);
}

vector<int> FindMinimumVisitsHelper(const vector<EndPoint>& endpoints) {
  vector<int> S;  // A minimum set of visit times.
  unordered_set<const Interval*> covered;
  vector<const Interval*> covering;
  for (const auto& e : endpoints) {
    if (e.is_left) {
      covering.emplace_back(e.ptr);
    } else if (covered.find(e.ptr) == covered.end()) {
      // e's interval has not been covered.
      S.emplace_back(e.ptr->right);
      // Adds all intervals in covering to covered.
      covered.insert(covering.cbegin(), covering.cend());
      covering.clear();  // e is contained in all intervals in covering.
    }
  }
  return S;
}
// @exclude

// O(n^2) checking solution.
void CheckAns(const vector<Interval>& I, const vector<int>& ans) {
  deque<bool> is_visited(I.size(), false);
  for (const int& a : ans) {
    for (size_t i = 0; i < I.size(); ++i) {
      if (a >= I[i].left && a <= I[i].right) {
        is_visited[i] = true;
      }
    }
  }

  for (const bool& b : is_visited) {
    assert(b == true);
  }
}

void SimpleTest() {
  vector<Interval> I;
  I.emplace_back(Interval{1, 4});
  I.emplace_back(Interval{2, 8});
  I.emplace_back(Interval{3, 6});
  I.emplace_back(Interval{3, 5});
  I.emplace_back(Interval{7, 10});
  I.emplace_back(Interval{9, 11});
  vector<int> ans = FindMinimumVisits(I);
  assert(ans.size() == 2 && ans[0] == 4 && ans[1] == 10);
}

int main(int argc, char* argv[]) {
  SimpleTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    cout << "Test " << times << endl;
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    vector<Interval> A;
    for (int i = 0; i < n; ++i) {
      uniform_int_distribution<int> left_dis(0, 9999);
      int left = left_dis(gen);
      uniform_int_distribution<int> right_dis(left, left + 100);
      int right = right_dis(gen);
      A.emplace_back(Interval{left, right});
    }
    vector<int> ans(FindMinimumVisits(A));
    CheckAns(A, ans);
  }
  return 0;
}

File: static/python/points_covering_intervals_alternative.py
import sys
import random


# @include
class Interval:

    def __init__(self, left=0, right=0):
        self.left = left
        self.right = right
# @exclude

    def __eq__(self, other):
        return self.left == other.left and self.right == other.right

    def __repr__(self):
        return '(%s, %s)' % (self.left, self.right)

    def __hash__(self):
        return self.left ^ self.right


# @include


class EndPoint:

    def __init__(self, interval, is_left):
        self.ptr = interval
        self.is_left = is_left

    def __lt__(self, other):
        a = self.ptr.left if self.is_left else self.ptr.right
        b = other.ptr.left if other.is_left else other.ptr.right
        return a < b or (a == b and self.is_left and not other.is_left)
# @exclude

    def __repr__(self):
        return '%s: (%s, %s)' % (
            ('Right', 'Left')[self.is_left], self.ptr.left, self.ptr.right)


# @include
def find_minimum_visits(intervals):
    endpoints = []
    for i in intervals:
        endpoints.append(EndPoint(i, True))
        endpoints.append(EndPoint(i, False))
    endpoints.sort()

    def find_minimum_visits_helper(endpoints):
        S = []  # A minimum set of visit times.
        covered = set()
        covering = []
        for e in endpoints:
            if e.is_left:
                covering.append(e.ptr)
            elif e.ptr not in covered:
                # e's interval has not been covered.
                S.append(e.ptr.right)
                # Adds all intervals in covering to covered.
                covered.update(covering)
                covering.clear()  # e is contained in all intervals in covering.
        return S
    return find_minimum_visits_helper(endpoints)
# @exclude


# O(n^2) checking solution
def check_ans(intervals, ans):
    is_visited = [False] * len(intervals)
    for a in ans:
        for i, interval in enumerate(intervals):
            if interval.left <= a <= interval.right:
                is_visited[i] = True
    assert all(is_visited)


def simple_test():
    intervals = [
        Interval(1, 4), Interval(2, 8), Interval(3, 6), Interval(3, 5),
        Interval(7, 10), Interval(9, 11)
    ]
    ans = find_minimum_visits(intervals)
    assert ans == [4, 10]
    intervals = [
        Interval(1, 2), Interval(2, 3), Interval(3, 4), Interval(4, 5),
        Interval(5, 6), Interval(6, 7)
    ]
    ans = find_minimum_visits(intervals)
    assert ans == [2, 4, 6]


def main():
    simple_test()
    for times in range(1000):
        print('Test', times)
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 10000)
        A = []
        for i in range(n):
            left = random.randrange(10000)
            right = random.randint(left, left + 100)
            A.append(Interval(left, right))
        ans = find_minimum_visits(A)
        check_ans(A, ans)


if __name__ == '__main__':
    main()
